### ISSUE TITLE
Frequent loss is NaN & Training Hangs

### ISSUE THREAD

**Issue by yikunzhang666** [aqlaboratory/openfold#50]
Thank you for sharing your code!

I am trying to train openfold, but the problem of loss being NAN persists, and the whole training hangs when this problem occurs.

I downloaded the code in early December and trained on 8 V100 cards with a training dataset size of 1000. When I ran to the 26th sample of the 2nd epoch, there were many warning outputs with a loss of NAN and the training was interrupted.
I read your solution of "Replace `training_step` in ` train_openfold.py` " in Issue #19, after changing, when I train the first sample, I got this:
```
WARNING:root:loss is NaN. Returning 0 loss...
```
Training still hangs.

I ran a recent commit again and retrained with the same dataset and the same problem occurred again and on the same sample. Like this:

![image](https://user-images.githubusercontent.com/52192467/148092680-7e734f59-8293-49ea-a515-21bf3f04a1d4.png)

I changed the way the mapping is generated in `data_modules.py` so that the dataset can be loaded in a fixed order when it is loaded, and I checked the data where the loss is NAN and found no abnormalities.

This is very strange, because with your first version of the code, there is no NAN loss so far, but with the version you committed after December this problem keeps occurring, even if I change my training dataset and the learning rate in the deepspeed config file, it does not improve the situation.

Is there a workaround for this situation?

**Comment by gahdritz** [comment]
What do you mean by "the version you committed after December"? Are you referring to a specific commit?

**Comment by ZwormZ** [comment]
Thanks for your quick response!

I downloaded your code three times, the first was the original version, the second was on December 6, and the third time was today when I downloaded the latest version of your commit. The initial version did not have the problem with the loss being NAN, but both subsequent versions did. 

I rerun the new version and there is the result :
![image](https://user-images.githubusercontent.com/52192467/148108151-6c19e7ce-8001-4ec9-afc3-213c7344751d.png)
This is the same situation as the version I downloaded on December 6 and ran with the same results.

**Comment by gahdritz** [comment]
The initial version didn't have the messages complaining about NaNs---is it possible that the losses were still going to NaN, just silently?

**Comment by ZwormZ** [comment]
Thanks for your reply!

About the iteration time, the model does run on GPUs, but according to the monitoring, when using deepspeed for multi-card training, the GPUs are occupied but the GPU memory is not always utilized,they will be utilized about 60% after 10 minutes interval, it seems that the training is stuck during this time. However, this did not happen when using one card for training, and the average time for each iteration was about 40s for single card training (1000 proteins were randomly selected in the pdb_mmcif file as the training set), so in my case the training speed with one card was about 1 times faster than training with 8 cards.

The details are shown in the figure below:
![image](https://user-images.githubusercontent.com/52192467/148181761-1115ab36-4c96-44ae-8b0c-deefebd98148.png)

I'm not sure if this is the cause, as it has been the case since my first training session, is this normal?

**Comment by gahdritz** [comment]
Kind of seems like you might be bottlenecked by data processing. Maybe try increasing the number of DataLoader workers?

**Comment by ZwormZ** [comment]
Thank you !

I tried move the training dataset to disk and set DataLoader workers to 16. The iteration time did get much shorter, it is now about 60s/it, but this is much longer than the 16s/it as you said, I want to test if it is the data that is causing the iterations to slow down.
It will be great if you can share with me a toy small data for the test you used in issue#34?

Also, I trained about 26 epochs on 1000 datasets, with 6 V100 cards, using the initial release code, and I drew the loss graph as follows:
![openfold_train](https://user-images.githubusercontent.com/52192467/148732659-8395c800-8b07-4de6-8da8-95206e067d6b.png)

It looks like the loss seems to stay around 3. and then not drop any further, did the loss also stay around 3. when you were training? I would like to know if this is normal.

**Comment by gahdritz** [comment]
Could you send the breakdown of the loss? Go to the definition of `AlphaFoldLoss` in `openfold/utils/loss` and print out the component parts of the cumulative loss.

**Comment by ZwormZ** [comment]
Sure, how can I send it to you conveniently? By email?

I will send you the original output loss detail after training 26 epochs, and I'm training again and print out the component parts of the cumulative loss, which may take some time.

**Comment by gahdritz** [comment]
Could you just post it here?

**Comment by ZwormZ** [comment]
[loss_1000_6cards.log](https://github.com/aqlaboratory/openfold/files/7843646/loss_1000_6cards.log)

**Comment by gahdritz** [comment]
Sorry if I was unclear, but I meant printing out the values of each of the constituent losses in `openfold/utils/loss.py` (e.g. FAPE loss, distogram loss, etc.). I want to see if one of the losses is dominating the final total of ~3.

**Comment by ZwormZ** [comment]
Sorry for the late reply, I spent some time retraining on two V100 cards , here is the loss breakdown.

[loss_1000_2cards.log](https://github.com/aqlaboratory/openfold/files/7848852/loss_1000_2cards.log)

 For visualization, I drew the constituent losses diagram for each part, as follows:
![distogram_loss](https://user-images.githubusercontent.com/52192467/148995544-5892ea77-c254-439a-bc01-9c2fe76b197f.png)
![fape_loss](https://user-images.githubusercontent.com/52192467/148995586-0bc1e123-fad7-4f66-a4f7-60556e1408e9.png)
![lddt_loss](https://user-images.githubusercontent.com/52192467/148995598-16f3a66b-032b-45ec-867f-58e01d445342.png)
![masked_msa_loss](https://user-images.githubusercontent.com/52192467/148995614-05dfea16-d86b-40ad-b378-54859958f018.png)
![supervised_chi_loss](https://user-images.githubusercontent.com/52192467/148995623-9e4e6b4a-f827-4912-be38-ed98fdc98802.png)
![cumulative_loss](https://user-images.githubusercontent.com/52192467/148995660-1100c875-030c-4302-8342-7b76e7552061.png)

**Comment by ZwormZ** [comment]
update：after ~4 epoch
![distogram_loss_4epoch](https://user-images.githubusercontent.com/52192467/149062880-4911432d-0ea6-4948-94a3-52639fab2497.png)
![fape_loss_4epoch](https://user-images.githubusercontent.com/52192467/149062882-694d3e98-822e-472c-b1c5-6edd81dd61a5.png)
![lddt_loss_4epoch](https://user-images.githubusercontent.com/52192467/149062884-73a62b47-37d5-46c1-9c5c-81dd22e0b9a1.png)
![masked_msa_loss_4epoch](https://user-images.githubusercontent.com/52192467/149062887-de22bd22-bbbb-4077-9c25-c6b217939151.png)
![supervised_chi_loss_4epoch](https://user-images.githubusercontent.com/52192467/149062890-3919fa33-81ee-48c3-9ab9-97e5f4d8b618.png)
![cumulative_loss_4eopch](https://user-images.githubusercontent.com/52192467/149062892-ea230408-9234-42c6-a0bd-02dc41b86b4f.png)
![loss_4epoch](https://user-images.githubusercontent.com/52192467/149062897-0b3f1d1b-718d-4cbd-a983-4ea17abd0dcf.png)

**Comment by ZwormZ** [comment]
I'd like to test if it's the data that causes the iteration times to be very different in my training process than in your case. It will be great if you can share with me a toy small data for the test you used in issue #34?

**Comment by empyriumz** [comment]
I got NaN loss in the training as well. I used `--precision 16` option, which might lead to overflow in this case, see https://github.com/pytorch/pytorch/issues/40497

A closer inspection suggests that in the `TriangleAttentionStartingNode` module where the inf is set to be `1e9`, `mask_bias = (self.inf * (mask - 1))[..., :, None, None, :]` yields `-inf`.

I tried to convert `mask` into fp32 and the `-inf` disappears, but no luck in converting it back to fp16.

BTW, the largest `self.inf` which fp16 can handle in my case in only `6e4`, not sure if it's safe or accurate to do so.

**Comment by gahdritz** [comment]
1e-9 is just the default---it should be overridden by the config file when you enable `--precision 16`. I've been focusing on bfloat16 training and implementing Multimer for the past couple of months, but I should have the bandwidth to the FP16 issues soon.

**Comment by empyriumz** [comment]
@gahdritz Thanks for the reply! It looks like the overflow is a persistent bug in the mixed precision training and maybe related to deepspeed as well.
I'll let you know if I can figure out the solution.
